smarts/core/smarts.py

```diff
@@ -614,34 +621,92 @@ def create_vehicle_in_providers(
                     # just use the first provider we find that accepts it
-                    ss = self._vehicle_index.sensor_state_for_vehicle_id(vehicle.id)
-                    provider.add_actor(state, ss.plan.route)
+                    # (note that the vehicle will already have a mission plan
+                    # registered for it in its sensor state in the vehicle_index.)
+                    provider.add_actor(state, prev_provider)
                     return
         # there should always be an AgentsProvider present, so we just assert here
         assert (
             False
         ), f"could not find a provider to accept vehicle {vehicle.id} for agent {agent_id} with role={role.name}"
 
-    def provider_relinquishing_vehicle(
-        self, state: VehicleState, cur_route: Optional[RoadMap.Route]
-    ):
-        """Find a new provider for a vehicle previously managed by an agent."""
+    def vehicle_exited_bubble(self, vehicle_id: str, teardown_agents: bool):
+        """Bubbles call this when a vehicle is exiting the bubble.
+        Will try to find a new provider for the vehicle if necessary."""
+        agent_id = None
+        shadow_agent_id = self._vehicle_index.shadow_actor_id_from_vehicle_id(
+            vehicle_id
+        )
+        if self._vehicle_index.vehicle_is_hijacked(vehicle_id):
+            agent_id = self._vehicle_index.actor_id_from_vehicle_id(vehicle_id)
+            self._log.debug(
+                f"agent={agent_id} relinquishing vehicle={vehicle_id} (shadow_agent={shadow_agent_id})"
+            )
+            state, route = self._vehicle_index.relinquish_agent_control(
+                self, vehicle_id
+            )
+            new_prov = self.agent_relinquishing_actor(
+                agent_id, state, teardown_agents, shadow_agent_id
+            )
+            if route and isinstance(new_prov, TrafficProvider):
+                new_prov.update_route_for_vehicle(vehicle_id, route)
+        elif shadow_agent_id:
+            self._log.debug(
+                f"shadow_agent={shadow_agent_id} will stop shadowing vehicle={vehicle_id}"
+            )
+            self._agent_manager.detach_sensors_from_vehicle(vehicle_id)
+            if teardown_agents:
+                self.teardown_agents_without_actors([shadow_agent_id])
+
+    def agent_relinquishing_actor(
+        self,
+        agent_id: str,
+        state: ActorState,
+        teardown_agents: bool,
+        shadow_agent_id: Optional[str] = None,
+    ) -> Optional[Provider]:
+        """Find a new provider for an actor previously managed by an agent.
+        Returns the new provider or None if a suitable one could not be found."""
+        provider = self._provider_for_actor(state.actor_id)
+        new_prov = self.provider_relinquishing_actor(provider, state)
+        if teardown_agents:
+            teardown_agent_ids = [agent_id]
+            if shadow_agent_id:
+                teardown_agent_ids.append(shadow_agent_id)
```

Comment:
I have some confusion here. The method seems to be dealing with a single agent but uses the plural. However, where I would expect that there might be plurals of `shadow_agent_id` there is just one.

Reply:
ah, yes, good point.  (I think that was left over from when I tried to do more than one at a time.)  Will fix to be singular.